src/Compiler/Postprocessor.php

```diff
@@ -30,26 +37,61 @@ final class Postprocessor
 
     private CompilerInterface $compiler;
 
+    private ScopeFactory $scopeFactory;
+
+    private NodeScopeResolver $nodeScopeResolver;
+
     public function __construct(
         Parser $parser,
         NodeVisitorStorage $nodeVisitorStorage,
         Standard $printerStandard,
-        CompilerInterface $compiler
+        CompilerInterface $compiler,
+        ScopeFactory $scopeFactory,
+        NodeScopeResolver $nodeScopeResolver
     ) {
         $this->parser = $parser;
         $this->nodeVisitorStorage = $nodeVisitorStorage;
         $this->printerStandard = $printerStandard;
         $this->compiler = $compiler;
+        $this->scopeFactory = $scopeFactory;
+        $this->nodeScopeResolver = $nodeScopeResolver;
     }
 
-    public function postProcess(string $phpContent, Template $template): string
+    public function postProcess(string $phpContent, Template $template, string $compileFilePath): string
     {
         $phpStmts = $this->findNodes($phpContent);
         foreach ($this->nodeVisitorStorage->getNodeVisitors() as $nodeVisitors) {
             $phpStmts = $this->processNodeVisitors($phpStmts, $nodeVisitors, $template);
         }
 
-        return $this->printerStandard->prettyPrintFile($phpStmts);
+        $phpContent = $this->printerStandard->prettyPrintFile($phpStmts);
+        file_put_contents($compileFilePath, $phpContent);
+        $realPath = realpath($compileFilePath) ?: '';
+        if ($realPath === '') {
+            return '';
+        }
+        require($compileFilePath); // load type definitions from compiled template
+
+        $scope = $this->scopeFactory->create(ScopeContext::create($realPath));
+
+        $phpStmts = $this->findNodes($phpContent);
+
+        $nodeScopeResolver = clone $this->nodeScopeResolver;
+        $nodeScopeResolver->processNodes($phpStmts, $scope, function (Node $node, Scope $scope) {
```

Comment:
What about moving this to ExprTypeNodeVisitorBehavior and run it on demand - when getType is called for first time - might remove need to do it in many templates at all

Reply:
See comment above